### PR TITLE
Secondary sales marketplace contracts

### PR DESCRIPTION
**Pull request by jamesmorgan**
### New Changes

* [x] Core functionality 
	* [x] Place bid
	* [ ] ~Increase bid~
	* [x] Withdraw bid
	* [x] Reject bid
	* [x] Accept bid
	* [x] Collab splits
	* [x] Check approval and transfers
* [x] Deploy to rinkeby
* [x] Deploy to mainnet
* [x] Tests passing

-----

* [x] Contracts - knownorigin/known-origin-contracts#3
* [x] Subgraph - knownorigin/known-origin-subgraph#6
* [x] Frontend - knownorigin/known-origin-web3-marketplace#209
* [x] API - https://github.com/knownorigin/known-origin-serverless-api/pull/45

### CODE REVIEW

File: contracts/v2/marketplace/TokenMarketplace.sol [hunk 1]
```diff
@@ -0,0 +1,344 @@
+pragma solidity ^0.4.24;
+
+import "openzeppelin-solidity/contracts/access/rbac/Roles.sol";
+import "openzeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "openzeppelin-solidity/contracts/access/Whitelist.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+interface ITokenMarketplace {
```
Comment (andygray) [inline contracts/v2/marketplace/TokenMarketplace.sol:8]: worth having in a separate file?

File: contracts/v2/marketplace/TokenMarketplace.sol [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+    // work out % of royalties total to split e.g. 43 / 85 = 50.5882353%
+    uint256 primaryArtistPercentage = _scaledUpCommission.div(_totalCollaboratorsRate);
+
+    uint256 totalPrimaryRoyaltiesToArtist = _remainingRoyalties.mul(primaryArtistPercentage).div(10 ** 18);
+    _artistAccount.transfer(totalPrimaryRoyaltiesToArtist);
+
+    uint256 remainingRoyaltiesToCollaborator = _remainingRoyalties.sub(totalPrimaryRoyaltiesToArtist);
+    _optionalCommissionRecipient.transfer(remainingRoyaltiesToCollaborator);
+  }
```
Comment (andygray) [inline contracts/v2/marketplace/TokenMarketplace.sol:289]: I am thinking this is sorta binding us only to offer one contributor max - or not allow secondary on multiple collabs (if we ever do them...)

File: contracts/v2/marketplace/TokenMarketplace.sol [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+  function setMinBidAmount(uint256 _minBidAmount) onlyIfWhitelisted(msg.sender) public {
+    minBidAmount = _minBidAmount;
+  }
+
+  function setKodavV2(IKODAV2 _kodaAddress) onlyIfWhitelisted(msg.sender) public {
+    kodaAddress = _kodaAddress;
+  }
+
```
Comment (andygray) [inline contracts/v2/marketplace/TokenMarketplace.sol:339]: Do we need setters for:

```  // 50 = 5%
  uint256 public defaultArtistRoyaltyPercentage = 50;

  // 30 = 3%
  uint256 public defaultKOPercentage = 30;```

File: contracts/v2/marketplace/TokenMarketplace.sol (hunk 1 above)
Comment (jamesmorgan) [reply contracts/v2/marketplace/TokenMarketplace.sol]: fixed

File: contracts/v2/marketplace/TokenMarketplace.sol (hunk 2 above)
Comment (jamesmorgan) [reply contracts/v2/marketplace/TokenMarketplace.sol:239]: Ye, we are bound as KODA v2 only supports two. And this is bound to KODA v2

File: contracts/v2/marketplace/TokenMarketplace.sol (hunk 3 above)
Comment (jamesmorgan) [reply contracts/v2/marketplace/TokenMarketplace.sol:302]: done

**Pull request merged**